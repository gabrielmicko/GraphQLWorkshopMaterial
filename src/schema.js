/**
 * Import makeExecutableSchema, resolvers, typeDefs
 * Call & export makeExecutableSchema, pass the imported variables
 */
import { makeExecutableSchema } from 'graphql-tools';
import resolvers from './resolvers';
import typeDefs from './types';

export default makeExecutableSchema({
  resolvers,
  typeDefs
});
