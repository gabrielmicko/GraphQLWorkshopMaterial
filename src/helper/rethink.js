import r from 'rethinkdb';
import speakers from '../db/speakers.json';
import talks from '../db/talks.json';

const connectToDB = () => {
  return r.connect({
    host: 'localhost',
    port: 28015,
    user: 'admin',
  });
};

const createDatabase = (conn, dbName) => {
  return new Promise(resolve => {
    r.dbCreate(dbName).run(conn, err => {
      err ? resolve(err) : resolve();
    });
  });
};

const createTable = (conn, tableName) => {
  return new Promise(resolve => {
    try {
      r.tableCreate(tableName).run(conn, err => {
        err ? resolve(err) : resolve();
      });
    } catch (e) {
      console.log(e);
    }
  });
};

const seedSpeakers = (conn, tableName) => {
  let insertPromises = [];
  speakers.forEach(speaker => {
    let insertPromise = new Promise((resolve, reject) => {
      r
        .db('jsdays')
        .table(tableName)
        .insert(speaker)
        .run(conn, (err, result) => {
          if (!err && result.inserted === 1) {
            resolve(result);
          } else {
            reject(err);
          }
        });
    });
    insertPromises.push(insertPromise);
  });
  return Promise.all(insertPromises);
};

const findSpeakersByName = (conn, tableName, name) => {
  return new Promise((resolve, reject) => {
    try {
      r.table(tableName).filter({ name: name }).run(conn, (err, res) => {
        if (err) {
          reject(err);
        } else {
          resolve(res.toArray());
        }
      });
    } catch (e) {
      console.log(e);
    }
  });
};

const seedTalks = (conn, talksTableName, speakersTableName) => {
  let insertPromises = [];
  talks.forEach(talk => {
    let talkInsertPromise = new Promise((resolve, reject) => {
      findSpeakersByName(conn, speakersTableName, talk.name).then(speakers => {
        if (speakers.length) {
          talk['speaker_id'] = speakers[0].id;
          r.table(talksTableName).insert(talk).run(conn, (err, result) => {
            if (!err && result.inserted === 1) {
              resolve(result);
            } else {
              reject(err);
            }
          });
        } else {
          resolve();
        }
      }, resolve);
    });
    insertPromises.push(talkInsertPromise);
  });
  return Promise.all(insertPromises);
};

const wipeTables = (conn, tables) => {
  let deletePromises = [];
  tables.forEach(table => {
    deletePromises.push(r.table(table).delete().run(conn));
  });
  return Promise.all(deletePromises);
};

export {
  connectToDB,
  createDatabase,
  createTable,
  seedSpeakers,
  seedTalks,
  wipeTables,
};
