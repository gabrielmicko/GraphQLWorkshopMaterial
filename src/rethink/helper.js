import r from 'rethinkdb';
import { RETHINK } from '../config/config.json';

/**
 * Creates a connection with the database.
 * The connection details are defined in this function.
 * @return {Promise<Object: conn>} Returns a Promise with a connnection
 */
const connectToDB = () => {
  return r.connect({
    host: RETHINK.HOST,
    port: RETHINK.PORT,
    user: RETHINK.USER
  });
};

/**
 * Creates a database according to the given argument.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @return {Promise} A promise which resolves when database created, rejects on any error
 */
const createDatabase = (conn, dbName) => {
  return new Promise(resolve => {
    r.dbCreate(dbName).run(conn, err => {
      err ? resolve(err) : resolve();
    });
  });
};

/**
 * Creates a table in the databse.
 * @param  {Object} conn Database connection
 * @param  {String} dbName The name of the database where the table is being created
 * @param  {String} tableName Name of the new table
 * @return {Promise} A promise which resolves when the table is created, rejects on any error
 */
const createTable = (conn, dbName, tableName) => {
  return new Promise(resolve => {
    r
      .db(dbName)
      .tableCreate(tableName)
      .run(conn, err => {
        err ? resolve(err) : resolve();
      });
  });
};

/**
 * Inserts data to the table. Checks if the dataArray argument is a valid array and its content is not empty.
 * Iterates over the dataArray and creates insert promises which are being collected.
 * Returns one promise which is fullyfied when all of the inserts are successful.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @param  {String} tableName Table name, where you insert data
 * @param  {Array} dataArray Array with Objects containing data to be inserted
 * @return {Promise} Promise which resolves when inserts are done.
 */
const insertData = (conn, dbName, tableName, dataArray) => {
  let insertPromisesArray = [];
  if (dataArray && Array.isArray(dataArray)) {
    dataArray.forEach(data => {
      let insertPromise = new Promise((resolve, reject) => {
        r
          .db(dbName)
          .table(tableName)
          .insert(data)
          .run(conn, (err, result) => {
            if (!err && result.inserted === 1) {
              resolve(result);
            } else {
              reject(err);
            }
          });
      });
      insertPromisesArray.push(insertPromise);
    });
  }
  return Promise.all(insertPromisesArray);
};

/**
 * Get all data from a table according to a filter object.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @param  {String} tableName Table name
 * @param  {Object} filters Object of containing the filter (ie. {speaker_id: 10}, check: toArray())
 * @return {Promise} Promise which resolves with the filtered data.
 */
const getByFilters = (conn, dbName, tableName, filters) => {
  return new Promise((resolve, reject) => {
    r
      .db(dbName)
      .table(tableName)
      .filter(filters)
      .run(conn, (err, result) => {
        !err ? resolve(result.toArray()) : reject(err);
      });
  });
};

/**
 * Get all the data from a table.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @param  {String} tableName Table name
 * @return {Promise} Promise which resolves with the data.
 */
const getAll = (conn, dbName, tableName) => {
  return new Promise((resolve, reject) => {
    r
      .db(dbName)
      .table(tableName)
      .run(conn, (err, result) => {
        !err ? resolve(result.toArray()) : reject(err);
      });
  });
};

/**
 * Deletes all the content from tables.
 * Iterates over the tables and creates delete promises which are being collected.
 * Returns one promise which is fullyfied when all of the deletes are successful.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @param  {Array} tables Array containing tables (strings), which you want to wipe (ie: ["speakers", "talks"])
 * @return {Promise} Promise which resolves when deletes are done.
 */
const wipeTables = (conn, dbName, tables) => {
  let deletePromises = [];
  tables.forEach(table => {
    let delPromise = new Promise(resolve => {
      r
        .db(dbName)
        .table(table)
        .delete()
        .run(conn, (conn, err) => {
          err ? resolve(err) : resolve();
        });
    });

    deletePromises.push(delPromise);
  });
  return Promise.all(deletePromises);
};

/**
 * Subscribe to any change that happens in a table.
 * Takes the callback function and calls it whenever something has changed.
 * The callback function in most of the cases should notify the PubSub system.
 * Returns one promise which is fullyfied when the subscription has been successful.
 * @param  {Object} conn Database connection
 * @param  {String} dbName Database name
 * @param  {Array} tableName Table name that you want to subscribe to.
 * @return {Promise} Promise which resolves when subscription is ready.
 */
const subscribeToTable = (conn, dbName, tableName, callbackFn) => {
  return new Promise(resolve => {
    r
      .db(dbName)
      .table(tableName)
      .changes()
      .run(conn, function(error, cursor) {
        if (error !== null) {
          reject(error);
        }
        cursor.each((e, c) => {
          if (c.new_val) {
            callbackFn(c.new_val);
          }
        });
        resolve();
      });
  });
};

export { connectToDB, createDatabase, createTable, insertData, wipeTables, getAll, getByFilters, subscribeToTable };
