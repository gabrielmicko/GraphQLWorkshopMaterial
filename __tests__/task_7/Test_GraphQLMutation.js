import { graphql } from 'graphql';
import schema from '../../src/schema';
import { connectToDB } from '../../src/rethink/helper.js';
import r from 'rethinkdb';

const deleteJohnDoe = conn => {
  return r
    .db('jsdays')
    .table('speakers')
    .filter({ name: 'John Doe' })
    .delete()
    .run(conn);
};

afterAll(() => {
  return new Promise(resolve => {
    connectToDB().then(conn => {
      deleteJohnDoe(conn).then(() => {
        conn.close();
        resolve();
      });
    });
  });
});

test('GraphQL Mutation addSpeaker.', () => {
  expect.assertions(1);
  const query = /* GraphQL */ `
    mutation addSpeaker {
      addSpeaker(
        name: "John Doe",
        bio: "John Doe is a test user.",
        facebook: "fb.com/john.doe",
        github: "github.com/john.doe",
        twitter: "github.com/john.doe",
        linkedin: "linkedin.com/john.doe",
        position: "Beeing a test everywhere.",
        photo: "/john.doe.png") {
          name,
          bio,
          facebook,
          github,
          twitter,
          linkedin,
          position,
          photo
        }
    }
  `;

  return new Promise(resolve => {
    graphql(schema, query).then(result => {
      resolve(expect(result).toMatchSnapshot());
    });
  });
});
