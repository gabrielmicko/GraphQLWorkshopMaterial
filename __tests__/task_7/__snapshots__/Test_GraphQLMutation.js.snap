// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Mutation addSpeaker. 1`] = `
Object {
  "data": Object {
    "addSpeaker": Object {
      "bio": "John Doe is a test user.",
      "facebook": "fb.com/john.doe",
      "github": "github.com/john.doe",
      "linkedin": "linkedin.com/john.doe",
      "name": "John Doe",
      "photo": "/john.doe.png",
      "position": "Beeing a test everywhere.",
      "twitter": "github.com/john.doe",
    },
  },
}
`;
